File: backend/app/schemas/task.py
"""Esquemas Pydantic de Tarea/Agenda."""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Estado = Literal["pendiente", "en_progreso", "completada"]
Prioridad = Literal["baja", "media", "alta"]


class TaskCreate(BaseModel):
    titulo: str = Field(min_length=3, max_length=160)
    descripcion: str | None = None
    fecha: datetime
    estado: Estado = "pendiente"
    prioridad: Prioridad = "media"


class TaskUpdate(BaseModel):
    titulo: str | None = Field(default=None, min_length=3, max_length=160)
    descripcion: str | None = None
    fecha: datetime | None = None
    estado: Estado | None = None
    prioridad: Prioridad | None = None
    completada: bool | None = None


class TaskOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    titulo: str
    descripcion: str | None = None
    fecha: datetime
    estado: Estado
    prioridad: Prioridad
    completada: bool
    usuario_id: int
    creado_en: datetime

File: gestor_agenda/lib/features/agenda/domain/entities/task.dart
/// Estados posibles de una tarea de la agenda.
enum EstadoTarea {
  pendiente('pendiente', 'Pendiente'),
  enProgreso('en_progreso', 'En progreso'),
  completada('completada', 'Completada');

  final String valor;
  final String etiqueta;
  const EstadoTarea(this.valor, this.etiqueta);

  static EstadoTarea desde(String valor) => EstadoTarea.values.firstWhere(
        (e) => e.valor == valor,
        orElse: () => EstadoTarea.pendiente,
      );
}

/// Nivel de prioridad de una tarea.
enum PrioridadTarea {
  baja('baja', 'Baja'),
  media('media', 'Media'),
  alta('alta', 'Alta');

  final String valor;
  final String etiqueta;
  const PrioridadTarea(this.valor, this.etiqueta);

  static PrioridadTarea desde(String valor) => PrioridadTarea.values.firstWhere(
        (p) => p.valor == valor,
        orElse: () => PrioridadTarea.media,
      );
}

/// Entidad de dominio: una actividad de la agenda.
class Task {
  final int id;
  final String titulo;
  final String? descripcion;
  final DateTime fecha;
  final EstadoTarea estado;
  final PrioridadTarea prioridad;
  final bool completada;

  const Task({
    required this.id,
    required this.titulo,
    this.descripcion,
    required this.fecha,
    required this.estado,
    required this.prioridad,
    required this.completada,
  });

  /// Una tarea esta vencida si su fecha ya paso y aun no se completa.
  bool get vencida =>
      !completada && fecha.isBefore(DateTime.now());
}
